kraken - ws - place orders

pcriadoperez opened the pull request:
- Add `createOrder`, `editOrder`, `cancelOrder`, `cancelOrders`, `cancelAllOrders`
- Add `timeInForce`

pcriadoperez commented:
### createOrder
```python
kraken.createOrder(BTC/USDT,limit,buy,0.0001,18000)
{'amount': 0.0001,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'fees': [],
 'filled': None,
 'id': 'O23SKT-6FN6Z-TQ2J6P',
 'info': {'descr': 'buy 0.00010000 XBTUSDT @ limit 18000.0',
          'event': 'addOrderStatus',
          'reqid': 1,
          'status': 'ok',
          'txid': 'O23SKT-6FN6Z-TQ2J6P'},
 'lastTradeTimestamp': None,
 'postOnly': None,
 'price': 18000.0,
 'reduceOnly': None,
 'remaining': None,
 'side': 'buy',
 'status': 'ok',
 'stopPrice': None,
 'symbol': 'BTC/USDT',
 'timeInForce': None,
 'timestamp': None,
 'trades': [],
 'triggerPrice': None,
 'type': 'limit'}
 ```
 ### editOrder
 ```python
 kraken.editOrder(O23SKT-6FN6Z-TQ2J6P,BTC/USDT,limit,buy,0.0001,18500)
{'amount': 0.0001,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'fees': [],
 'filled': None,
 'id': 'OKW3CQ-KWPHC-WYNYJO',
 'info': {'descr': 'buy 0.00010000 XBTUSDT @ limit 18500.0',
          'event': 'editOrderStatus',
          'originaltxid': 'O23SKT-6FN6Z-TQ2J6P',
          'reqid': 1,
          'status': 'ok',
          'txid': 'OKW3CQ-KWPHC-WYNYJO'},
 'lastTradeTimestamp': None,
 'postOnly': None,
 'price': 18500.0,
 'reduceOnly': None,
 'remaining': None,
 'side': 'buy',
 'status': 'ok',
 'stopPrice': None,
 'symbol': 'BTC/USDT',
 'timeInForce': None,
 'timestamp': None,
 'trades': [],
 'triggerPrice': None,
 'type': 'limit'}
 ```
 ### cancelOrder
 ```python
 kraken.cancelOrder(OKW3CQ-KWPHC-WYNYJO)
{'event': 'cancelOrderStatus', 'reqid': 1, 'status': 'ok'}
```
### cancelAllOrders
```python
kraken.cancelAllOrders()
{'count': 2, 'event': 'cancelAllStatus', 'reqid': 1, 'status': 'ok'}
```

sc0Vu commented on js/pro/kraken.js:
```diff
@@ -57,6 +63,197 @@ module.exports = class kraken extends krakenRest {
+    async createOrder (symbol, type, side, amount, price = undefined, params = {}) {
+        /**
+         * @method
+         * @name kraken#createOrder
```
Shall we add doc url for `createOrder`, `cancelOrder` and `cancelAllOrders`?

carlosmiei commented:
I think this is the natural evolution for ccxt, but do we have a consensus yet?

Personally, I'm not sure yet if you should be overriding the Rest implementation because it might be a surprise for some users since the raw response might be different, RL might be different as well because don't have that implemented properly in the WS side, and the error handling might not be prepared yet to handle all order manipulation WS errors. 

@pcriadoperez @kroitor @frosty00, what do you think?

pcriadoperez commented:
> I think this is the natural evolution for ccxt, but do we have a consensus yet?
> 
> Personally, I'm not sure yet if you should be overriding the Rest implementation because it might be a surprise for some users since the raw response might be different, RL might be different as well because don't have that implemented properly in the WS side, and the error handling might not be prepared yet to handle all order manipulation WS errors.
> 
> @pcriadoperez @kroitor @frosty00, what do you think?

Hi, I understood we agreed on this on discord. But maybe we should wait for @frosty00 implementation of the binance ws api he is working on to make sure it works in the same way.

About the override, on one side I think it's ok for users to assume that if they use the pro exchange we will default to websockets. However I agree it could be a breaking change for anyone using the raw output, I think the question here would be, would defaulting to using websockets benefit more users than any possible breaking changes? I could agree having some sort of flag or warning, or being able to choose if to use the rest api vs ws api through a parameter or option.

About the errors do you mean because we don't have the exceptions as well documented on the WS side?

ttodua commented:
Personally I, would vote against override. It has more potential damage. What will hurt if the name was different? i.e. `createWsOrder` (or `createOrderWs`) ? would be win-win for anyone. Overriding it definitely carries some actual risks, like:
- REST endpoint (engine) could be maintained by different team than WS engine, or their Docs & Development was still independent from REST (even if same person works in the backend) , thus REST endpoints might be supporting more things (i.e. `stop-loss, ioc, whatever...` while WS endpoints might not be supporting them. thus overwriting `createOrder` is regress.
- overriding REST endpoints by WS is still not ideal, because some people running `createOrder` might have some technical/backend reasons to do it through REST call (there could valid cases for that) and didn't want to have mandatory WS call for submitting orders. Don't forget, there is not solely JS devs, but PHP/PY use-cases might be different, so we should leave REST methods. 
- in the end, we would have a clearcut, distinctive design of method names, as opposed so mentioning method name is enough to know what is being referred to (without need to ask that person "hey, did you use WS 'createOrder' or REST 'createOrder' ?)
- might be some other potential disadvantage I cant think of atm, but...  what are its disadvantages of having  `createOrderWs`? lets put them on second plate to compare.

pcriadoperez commented:
I thought I would bring this back to surface as I think there are some more exchanges that allow trading using websockets. I updated it to use typescript and changed the method names so they don't overright the rest ones and added the errors listed in kraken.

carlosmiei commented on ts/src/kraken.ts:
```diff
@@ -1528,6 +1546,10 @@ export default class kraken extends Exchange {
             request['close'] = close;
         }
+        const timeInForce = this.safeString2 (params, 'timeInForce', 'timeinforce');
+        if (timeInForce !== undefined) {
```
we're missing omitting 'timeInForce' here
